File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py


def solution(numbers, target):

    def dfs(idx, num):
        nonlocal cnt
        if idx == len(numbers):
            if num == target :
                cnt += 1
            return 
        dfs(idx+1, num+numbers[idx])
        dfs(idx+1, num-numbers[idx])
    
    cnt = 0
    dfs(0,0)

    return cnt

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int answer =0;
    public int solution(int[] numbers, int target) {
        dfs(numbers, 0,0,target);
    
        return answer;
    }
    
    public void dfs(int[] numbers, int depth, int sum, int target){
        //마지막 노드까지 탐색한 경우
        if(depth == numbers.length){
            if(target == sum) answer++ ;
            return;
        }
        int plus = sum + numbers[depth];
        int minus = sum - numbers[depth];
        
        dfs(numbers,depth+1, plus, target);
        dfs(numbers, depth+1, minus, target);
    }
}